### PR TITLE
Bubbles: avoid redirect by adjusting protocol

### PR DESCRIPTION
**Pull request by breville**
Bubbles above a level have had `http://` links (from [here](https://github.com/code-dot-org/code-dot-org/blob/7a17ac510c9db89cbc35ac8f657a4b92401b58a3/dashboard/app/helpers/levels_helper.rb#L49)), even when viewed using a `https://` connection, leading to a redirect upon every bubble click.

Now, the bubble links are a protocol-relative `//` so that the links match the current protocol.

### Impact on the Maker App

The Code.org Maker App, source [here](https://github.com/code-dot-org/browser), has had an issue in which a Circuit Playground board isn't accessible after navigating to a level using a bubble.  It appears to be this [issue](https://github.com/electron/electron/issues/4025), and rather than changing the Maker App right now, this should work around the issue.

### CODE REVIEW

File: apps/src/templates/progress/BubbleFactory.jsx
Comment (davidsbailey) [inline apps/src/templates/progress/BubbleFactory.jsx]: naively (without understanding the whole call stack), I would ask why it isn't possible to have levelUrl be a protocol-relative url here so that we don't have to rewrite it?

**Review by davidsbailey — COMMENTED** [review]
this part of the code is owned by @code-dot-org/learning-platform so I would defer to their judgment here.

File: apps/src/templates/progress/BubbleFactory.jsx
Comment (jamescodeorg) [reply apps/src/templates/progress/BubbleFactory.jsx]: I think this url might be coming from the back-end (possibly [here](https://github.com/code-dot-org/code-dot-org/blob/7981d80e8bc7a84e7902658872c5adadc141e32f/dashboard/app/models/script_level.rb#L326)?).  If so, maybe we should change this on the back-end to use the correct protocol or make it relative as Dave suggested?

Comment (breville) [reply apps/src/templates/progress/BubbleFactory.jsx]: I believe the URL comes from [here](https://github.com/code-dot-org/code-dot-org/blob/7a17ac510c9db89cbc35ac8f657a4b92401b58a3/dashboard/app/helpers/levels_helper.rb#L49).  I am happy to try making this URL protocol-relative, but I was a bit nervous about the impact of such a change.

Comment (jamescodeorg) [reply apps/src/templates/progress/BubbleFactory.jsx]: Are we able to make this URL always match the scheme of the original request?

Comment (breville) [reply apps/src/templates/progress/BubbleFactory.jsx]: Passing the request's protocol from the controller through to the model and then helper seems a bit heavy, but using a protocol-relative URL seems to have worked.

Comment (jamescodeorg) [reply apps/src/templates/progress/BubbleFactory.jsx]: Sorry for belaboring this...  I actually didn't know that you could have an url that was just relative to the protocol (and not protocol+host+port) and did some reading on it and at least [some people](https://www.paulirish.com/2010/the-protocol-relative-url/) recommend against this approach.  Given our constrained use, I think it might still be ok but I wanted to ask -- what if we just set the protocol to `https`?  It sounds like we typically enforce this in our production environments (see slack thread in #infrastructure).  Would this break our test/dev environments?

Comment (davidsbailey) [reply apps/src/templates/progress/BubbleFactory.jsx]: great point James, I was not aware that protocol-relative urls were problematic. unfortunately, sending us to https will break things on localhost and drone.

I think what I am missing is why we are using a full url (including protocol and hostname) instead of just using the path directly. do we expect these bubbles to work from a context outside of dashboard server?

here are the relevant rails docs: https://guides.rubyonrails.org/routing.html#path-and-url-helpers here is what I see on the staging machine:
```
[staging] dashboard > include Rails.application.routes.url_helpers
=> Object
[staging] dashboard > root_url
=> "http://staging-studio.code.org/"
[staging] dashboard > root_path
=> "/"
```
I started to look into whether there is a railsy way to make calls like root_url preserve the protocol, but that seems secondary to why we are including the protocol and hostname in the first place.

Comment (davidsbailey) [reply apps/src/templates/progress/BubbleFactory.jsx]: > I think what I am missing is why we are using a full url (including protocol and hostname) instead of just using the path directly. do we expect these bubbles to work from a context outside of dashboard server?

if we want to avoid this question, we could also use `CDO.studio_url` as a drop-in replacement for `url_from_path`.

Comment (davidsbailey) [reply apps/src/templates/progress/BubbleFactory.jsx]: actually, `CDO.studio_url` returns a protocol-relative url!
```
[staging] dashboard > CDO.studio_url('foo')
=> "//staging-studio.code.org/foo"
```
that means we are already using protocol-relative urls quite widely in our application. so I would suggest we stick with allowing them in this PR.

I think it would be worth following up to make `url_from_path` use `CDO.studio_url` used instead of `root_url(protocol: ...)`, since `CDO.studio_url` is by far more common in our codebase, and there are only a couple of instances of `root_url` to eliminate.

James, if you want to push to eliminate protocol-relative urls, that could be a bigger change that might be best to take through whatever process we have for dealing with security issues (e.g. posting in #infrastructure).

Comment (davidsbailey) [reply apps/src/templates/progress/BubbleFactory.jsx]: > James, if you want to push to eliminate protocol-relative urls, that could be a bigger change that might be best to take through whatever process we have for dealing with security issues (e.g. posting in #infrastructure)

looking more... given that we already redirect all http requests to https in production, it's hard to imagine how protocol-relative urls could be hurting us. I would expect these only to hurt us if it meant that requesting a page in http allowed an attacker to force that page to load a bunch of assets also in http.

File: dashboard/test/models/script_level_test.rb
Comment (davidsbailey) [inline dashboard/test/models/script_level_test.rb]: I would suggest using `CDO.studio_url` here too, and below:
```
[staging] dashboard > root_url(protocol: '//').chomp('/')
=> "//staging-studio.code.org"
[staging] dashboard > CDO.studio_url
=> "//staging-studio.code.org"
```

```suggestion
    assert_match Regexp.new("^#{CDO.studio_url}/s/bogus-script-[0-9]+/lessons/1/levels/1$"), summary[:url]
```

**Review by davidsbailey — APPROVED** [review]
LGTM after addressing one more comment. thanks for helping clean this up!

File: dashboard/test/models/script_level_test.rb
Comment (breville) [reply dashboard/test/models/script_level_test.rb]: Oh great catch!

**Pull request merged**
